[REPLACED WITH SPLITTED STACK] Cache kernel code generation in TritonTemplate.generate()

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #149267


In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850 ->20890757277
33527024884- > 28316151238

another win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
cache kernel codegen and loading

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "cache kernel codegen and loading"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Split cashable portion of generate into its own function."



This step is a prepration  step for the next PR where i will add a cache on top of generate_and_load.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Split cashable portion of generate into its own function."



This step is a prepration  step for the next PR where i will add a cache on top of generate_and_load.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Split cashable portion of generate into its own function."



This step is a prepration  step for the next PR where i will add a cache on top of generate_and_load.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Split cashable portion of generate into its own function."



This step is a prepration  step for the next PR where i will add a cache on top of generate_and_load.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Split cashable portion of generate into its own function."



This step is a prepration  step for the next PR where i will add a cache on top of generate_and_load.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see 30% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the modules.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see 30% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the modules.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see 30% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the modules.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+        return (
+            entry.mod,
+            entry.extra,
+            tuple(
```
Maybe note that you're not just returning the entry directly so you can map the entry to the input names of the current object...

aorenste commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1059,50 +1191,49 @@ def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
+        clear_on_fresh_inductor_cache(self._generated_module_cache)
 
-    def generate(  # type: ignore[override]
+    # Those class fields are used for testing _generated_module_cache.
```
"Those" -> "These"

aorenste commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1059,50 +1191,49 @@ def __init__(self, name: str, grid: Any, source: str, debug=False) -> None:
+    # When this flag is on, we ensure that the cached results and the generated result if cache
+    # was not used are the same.
+    test_cache = False
+    generated_module_cache_hit = 0
```
Might be useful to have cache misses as well. (10 hits - yay - is that good? Out of 10 attempts? Yes.  Out of 100k? No.)

aorenste commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1134,6 +1262,17 @@ def generate(  # type: ignore[override]
             "subgraphs": subgraphs,
         }
 
+        cache_result = self._generated_module_cache.get_entry(cache_key, input_nodes)
```
Should there be a way to disable the cache?

laithsakka replied:
yeh probably let me add that

aorenste commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1134,6 +1262,17 @@ def generate(  # type: ignore[override]
+        cache_result = self._generated_module_cache.get_entry(cache_key, input_nodes)
+
+        if cache_result is not None:
+            TritonTemplate.generated_module_cache_hit = (
+                TritonTemplate.generated_module_cache_hit + 1
+            )
```
```suggestion
            TritonTemplate.generated_module_cache_hit += 1
```

aorenste reviewed (commented):
Looks pretty good to me - a few small comments.

bobrenjc93 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+    # (x, y, x) we generate {x->0, y->1, x->2}. We also generate the inverse mapping {2->x, 1->y}.
+
+    # After running the codegen, for the outputs kernel.args.input_buffers.keys() and
+    # if we do cache the result we cache the indices of the inputs instrea.
```
lol i feel like I'm a neural network trained on Laith's brain. I think he meant "instead".

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+    code: str
+
+
+class GeneratedModulesCache(dict[str, GeneratedModulesCacheEntry]):
```
i need the cache_clear to be able to register to be cleared with fresh_inductor_cache()
but I can remove the inheritance

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+
+    # To address that, we do normalization using input_name_to_index and index_to_input_name:
+    # Each input name is mapped to its index in the inputs list, for example for inputs
+    # (x, y, x) we generate {x->0, y->1, x->2}. We also generate the inverse mapping {2->x, 1->y}.
```
i listed x on purpose to show it work for this corner case as well.
I generate the mapping from the input_list so it would actually be
 {y->1, x->2}.
 and
 {0->x, 1->y, x->2}
 i think i i reversed the arrow above lol!

laithsakka replied:
fixed

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+            and call_sizes == layout.size
+            and workspace_arg is None
+        ):
+            cache_key = repr(
```
i saw that used in other cache key computations

aorenste replied:
It probably works - it's just a bit odd to do stringy comparisons which may be unexpectedly lossy.

laithsakka replied:
I can use   if its safer   return pickle.dumps(cache_key) but i need to return a hashable,

bobrenjc93 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1050,6 +1051,137 @@ def _jinja2_env():
+    # if this is called again and we get a cache hit with another inputs (m, h, m) then we can
+    # use that indexing to regenerate the kernel.args.input_buffers.keys() it would be (m, h, m).
+
+    # TODO Normalize symbols in input_nodes so that inputs
```
nit please file an issue and reference the issue number in the TODO

eellison reviewed (commented):
within lowering, we already have a precompile cache: 

https://github.com/pytorch/pytorch/blob/27370998b2b327fe5e8eaf84d6f83f6349847289/torch/_inductor/select_algorithm.py#L1649

And the choice callers already have a hash key. 

What about something where we check before rendering if there is an existing rendered TritonTemplate from the precompile cache key and grab that rendered string ?

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation and loading in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
do not cache the generation of the python code and the laoding for each config that we autotune on.
This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
profile:
<img width="726" alt="Screenshot 2025-03-20 at 1 23 25 PM" src="https://github.com/user-attachments/assets/9c94ba57-fc84-42c5-9ef5-3ed5da26f738" />

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "[Draft] Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1199,22 +1340,166 @@ def generate(  # type: ignore[override]
+            kernel_args_sizevars_keys,
+            kernel_options,
+        )
+
```
this is existing code moved. appear here as new lol

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850->23319257037
33527024884->31341716252

A bigger win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850->23319257037
33527024884->31341716252

A bigger win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850->23319257037
33527024884->31341716252

A bigger win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
added unit test diff is ready for review

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850->23319257037
33527024884->31341716252

A bigger win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison reviewed (commented):
Cool! Before I review - do we have a sense of gain on the model ? I see 20% on this benchmark, which is nice..

laithsakka commented:
> Cool! Before I review - do we have a sense of gain on the model ? I see 20% on this benchmark, which is nice..

yeh this is around 13% of the total model compile time.

eellison commented on test/inductor/test_max_autotune.py:
```diff
@@ -1128,6 +1133,141 @@ def f(x, y):
+            TritonTemplate.generated_module_cache_hit = 0
+            TritonTemplate.generated_module_cache_miss = 0
+
+        def func_test1(x, y, z, m):
```
nit: set `max_autotune_gemm_backends = "TRITON"` to ensure we use this tritons. Also set `max_mm_configs` to a lower number to speed up tests.

laithsakka replied:
done

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
         self.template_out: Optional[str] = None
         self.ops_handler: Optional[V.WrapperHandler] = None  # type: ignore[name-defined]
 
+        # The generated code is not dependent on the input nodes names, or symbolic sizes names.
```
This is just for the template generation of non-fused templates. It wont be true for templates which undergo fusion. I'm not sure exactly how we're avoiding that today. Could we more explicitly opt out of caching for that case ?

We only attempt to fuse epilogues onto one template of the original set by default, so its not worth caching anyway. But we might attempt more. You can test this with adding a `relu` at the end of an mm and backends=TRITON i mentioned above.

This is the code path for it https://github.com/pytorch/pytorch/blob/5a422150c32f2c861061dc8be5f8d19fb6d80155/torch/_inductor/codegen/simd.py#L1461-L1471

eellison replied:
We should also skip if subgraphs exist. Or for flexattention / subgraph_bodies..

laithsakka replied:
this code path is not on the cache path. i.e: it does not call generate_and_load since it direcly call render
that said we are redundantly recording the state:
here is what I will do to address there.

1) as mentioned in a different comment, I will make this opt in so we control locations that actually uses the caching .
2) I will update PR so that the recording does not happen when not needed.

laithsakka replied:
addressed as explained.

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
+    def input_dependent_preserved_state(self) -> str:
+        return repr(
+            [
+                self.args.input_buffers,
```
Can we just return the full repr for `self.args` ? Some templates use a workspace arg:

https://github.com/pytorch/pytorch/blob/5a422150c32f2c861061dc8be5f8d19fb6d80155/torch/_inductor/select_algorithm.py#L337-L340

laithsakka replied:
this will always end up caching ops what we do not always need and effect perf significantly.

laithsakka replied:
added  self.args.workspace_args

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
+        return repr(
+            [
+                self.args.input_buffers,
+                self.args.sizevars,
```
nit: can we include this list above, in __init__, so it will be more visible to people who add more attrs ?

laithsakka replied:
how can we do that? like in the comment?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
         self.ops_handler: Optional[V.WrapperHandler] = None  # type: ignore[name-defined]
 
+        # The generated code is not dependent on the input nodes names, or symbolic sizes names.
+        # However, some of the variables returned by generate_and_load that are computed during the
```
Can we have something louder ? then full list

XXX: NEW FIELDS WHICH ARE MODIFIED DURING KERNEL DEFINITION (not including fusion) MUST BE ADDED HERE

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
+            ]
+        )
+
+    def record_input_dependent_tracked_event(self) -> Callable[..., Any]:
+        def decorator(fn) -> Callable[..., Any]:
+            def wrapper(*args, **kwargs) -> Any:
+                pre_state = self.input_dependent_preserved_state()
+                result = fn(*args, **kwargs)
+                post_state = self.input_dependent_preserved_state()
+                if pre_state != post_state:
+                    self.cached_replay_events.append((fn.__name__, [*args], {**kwargs}))
+                return result
```
Cool !

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+        symbol_mapping = {}
+        next_symbol_index = 1
+
+        def normalize_symbols(x: sympy.core.Expr) -> str:
+            symbols = x.free_symbols
+            nonlocal next_symbol_index
+            original_string = str(x)
+            for s in symbols:
+                if str(s) not in symbol_mapping:
+                    symbol_mapping[str(s)] = f"_normalized_symbol{next_symbol_index}"
+                    next_symbol_index += 1
+                original_string = original_string.replace(
+                    str(s), symbol_mapping[str(s)]
+                )
```
Can we include this in a separate PR ? or i need to understand better the problem here.. but separate pr might be better

laithsakka replied:
after I address all the comments I will see how easy that would be . it will be a alot of changes in the unit test also

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+                ]
+            )
+
+        if subgraphs is None and workspace_arg is None:
```
would not need the workspace arg carve out if we just included full args

laithsakka replied:
I added the workspace arg in the state, is this needed for mm_template?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+                    "num_consumer_groups": num_consumer_groups,
+                    "num_buffers_warp_spec": num_buffers_warp_spec,
+                    "kwargs": kwargs,
+                    "epilogue_fn": id(epilogue_fn),
```
hmm, while I don't think we will have non persistent epilogue_fns... is it worth keeping around somewhere ? or just doing full repro of the fn ?

laithsakka replied:
what do you mean by keeping around? I think initially we can just cache if its none.

laithsakka replied:
addressed as explained

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1198,22 +1346,166 @@ def generate(  # type: ignore[override]
                 )
-
             extra = "-".join(extra_parts) + "-"
-            mod = PyCodeCache.load(code, extra)
```
why do we need to do this ?

laithsakka replied:
this function is make_extra is just computing extra and returning it.
unrelated but not sure if this what you are asking about. 
mod also is not cached so its moved outside the caching lookup block. (it has its own cache)

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1287,7 +1578,7 @@ def make_kernel_render(out_node):
             mutated_inputs=mutated_inputs,
             workspace_arg=workspace_arg,
-            allowed_prologue_inps=kernel.prologue_supported_inputs.copy(),
+            allowed_prologue_inps=prologue_supported_inputs,
```
Can we keep the clone for now ? i had some concern around sharing mutable state..

eellison replied:
oh its done above..

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+        tuple[ModuleType, str, tuple[str, ...], OrderedSet[str], Any, dict[str, Any]]
+    ]:
+        """Generate the python code and load it into the current process"""
+        cache_key = self._generated_code_cache.make_key(
```
This part is rather hard to review.. would it be possible to stack it, and in the bottom pr always cache miss ?

laithsakka replied:
I am not sure how easy that would be at this moment i will loop at splitting at the end after i address all comments

eellison commented on test/inductor/test_max_autotune.py:
```diff
@@ -1128,6 +1133,141 @@ def f(x, y):
+
+        # Test duck typing.
+        with fresh_inductor_cache():
+            reset_counters()
```
Can we check results vs eager ?

laithsakka replied:
done

laithsakka replied:
actually eager and compile are not the same even with the cache being disabled so am not sure ... how do you test equality is there acceptable error margin?

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1066,63 +1113,177 @@ def _jinja2_env():
+            return [normalize_symbols(x) for x in ls]
+
+        def layout_key(layout: ir.Layout) -> str:
+            return repr(
```
can we assert that this is not a FlexibleLayout?

laithsakka replied:
done we will not cache if any input or the output is FlexibleLayout

eellison reviewed (commented):
Looking v good ! left some comments.

Can we do the symints in the next pr? also, if possible, stack pr below where we always cache miss..

drisspg reviewed (commented):
I am still reading through this, but my initial comment is that I think that this code gen caching should be opt-in since. IMO it feels like the caching will be hard to get completely right and knowing when a newly added template (or existing one like flex) is safe to be cached seems error prone

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -366,6 +372,46 @@ def __init__(
+        # However, some of the variables returned by generate_and_load that are computed during the
+        # triton template expansions (code generation) are dependent on those.
+        # In order to cache the code generation and avoid redoing it for similar inputs that varies only by
+        # input names or symbol names, we do a record and replay method.
```
as @drisspg pointed out, we also need to be careful about global/input state. for instance, these freezing of layouts:

https://github.com/pytorch/pytorch/blob/6fb089f2a2eea75a45ac2340f0e68736524e20bf/torch/_inductor/select_algorithm.py#L922-L927

laithsakka replied:
I am unsure how to capture this, maybe add a counter to the numbers of freezed inputs and add to the state?
but we have to capture all such global input mutations  
what do you guys think?
in addition to that i think it make sense to make this opt in. I will make it opt-in in two levels. 

  1. opt in with respect to the template (make the mm_template  only for now opt into this)
  2. opt in as not enabled by default also.
  
 
 what do you guys think
 @eellison @drisspg

laithsakka replied:
addressed as explained,

laithsakka commented:
> I am still reading through this, but my initial comment is that I think that this code gen caching should be opt-in since. IMO it feels like the caching will be hard to get completely right and knowing when a newly added template (or existing one like flex) is safe to be cached seems error prone

I can make it opt in, i think its mostly needed for mm template @eellison  what do you think?

laithsakka commented on test/inductor/test_max_autotune.py:
```diff
@@ -1128,6 +1133,141 @@ def f(x, y):
+    @config.patch(max_autotune=True)
+    def test_triton_template_generated_code_cache(self):
+        def reset_counters():
+            TritonTemplate.generated_module_cache_hit = 0
```
done

etaf commented on test/inductor/test_max_autotune.py:
```diff
@@ -1128,6 +1133,141 @@ def f(x, y):
+            b = torch.matmul(z, m)
+            return a, b
+
+        a = torch.rand(10, 22, device="cuda")
```
Hi, May I suggest we replace `"cuda"` with `GPU_TYPE` here? This case also runs on XPU, the hard code `"cuda"` here will fail on XPU. Thanks.

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850 ->20890757277
33527024884- > 28316151238

another win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850 ->20890757277
33527024884- > 28316151238

another win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
address all comments except splitting the PR. i will address publish the split in a seperate stack and link it here.

laithsakka committed:
Update on "Cache kernel code generation in TritonTemplate.generate() "



In a model, we see ~~ 40% of the time in mm/addmm tuning. The model have 2000 mm,
many of which receives the same input shapes. 

with autotune enabled, this become expensive, while we already cache auto tuning results, we
did not used to cache the generation of the python code and the loading for each config that we autotune on.

This diff handles the code generation part (template expansions) a previous diff handled the loading part. 
This is expected to save 20% of the model I am working on. 

How do we do the caching?
For a given configurations and input layout, the generated code is always the same. One caveat is that 
some other information collected during code generation are input dependent (namely depends on inputs
names and symbol names in inputs). and not just layout. ! 
To handle those we use a record and replay approach, where we record the functions that are called during 
code generation that effect those outputs and replay them at a cache hit. 

effect on mm_loop benchmark :
26262641850 ->20890757277
33527024884- > 28316151238

another win will come on the next PR on top of this, when we avoid replaying load_input.
by changing the way kernel.prologue_supported_inputs are computed. 

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
I am splitting this no need to review this except for reading my responses.

laithsakka commented:
I addressed all comments and splitted the PR see the new stack at https://github.com/pytorch/pytorch/pull/151764
can you take another look ? @eellison @drisspg
